File: implementations/java/src/main/java/com/webweavex/Pipeline.java
package com.webweavex;

import java.util.*;

public class Pipeline {
    private Cleaner cleaner;
    private Chunker chunker;
    private Entities entities;
    private Relations relations;
    private Graph graph;
    private Insights insights;
    
    public Pipeline() {
        this.cleaner = new Cleaner();
        this.chunker = new Chunker();
        this.entities = new Entities();
        this.relations = new Relations();
        this.graph = new Graph();
        this.insights = new Insights();
    }
    
    public Map<String, Object> extractFromText(String text) {
        String cleanedText = cleaner.clean(text);
        List<Map<String, Object>> chunks = chunker.chunk(cleanedText);
        List<Map<String, String>> entityList = entities.extract(cleanedText);
        List<Map<String, String>> relList = relations.extract(entityList);
        Map<String, Object> graphResult = graph.build(entityList);
        Map<String, Object> insightsResult = insights.compute(entityList, chunks, cleanedText);
        
        return buildResult("", "", cleanedText, chunks, entityList, relList, graphResult, insightsResult);
    }
    
    private Map<String, Object> buildResult(String url, String title, String text,
                                            List<Map<String, Object>> chunks,
                                            List<Map<String, String>> entityList,
                                            List<Map<String, String>> relationsList,
                                            Map<String, Object> graphResult,
                                            Map<String, Object> insightsResult) {
        Map<String, Object> result = new LinkedHashMap<>();
        
        Map<String, String> meta = new LinkedHashMap<>();
        meta.put("title", title);
        meta.put("url", url);
        result.put("meta", meta);
        
        Map<String, String> content = new LinkedHashMap<>();
        content.put("text", text);
        result.put("content", content);
        
        List<Map<String, Object>> sortedChunks = new ArrayList<>(chunks);
        sortedChunks.sort((a, b) -> ((Integer) a.get("index")).compareTo((Integer) b.get("index")));
        result.put("chunks", sortedChunks);
        
        List<Map<String, String>> sortedEntities = new ArrayList<>(entityList);
        sortedEntities.sort((a, b) -> {
            int typeComp = a.get("type").compareTo(b.get("type"));
            if (typeComp != 0) return typeComp;
            return a.get("value").compareTo(b.get("value"));
        });
        result.put("entities", sortedEntities);
        
        List<Map<String, String>> sortedRelations = new ArrayList<>(relationsList);
        sortedRelations.sort((a, b) -> {
            int sourceComp = a.get("source").compareTo(b.get("source"));
            if (sourceComp != 0) return sourceComp;
            return a.get("target").compareTo(b.get("target"));
        });
        result.put("relations", sortedRelations);
        
        result.put("graph", graphResult);
        result.put("insights", insightsResult);
        
        return result;
    }
}

File: implementations/kotlin/src/main/kotlin/com/webweavex/Pipeline.kt
package com.webweavex

import com.google.gson.GsonBuilder

class Pipeline {
    private val cleaner = Cleaner()
    private val chunker = Chunker()
    private val entities = Entities()
    private val relations = Relations()
    private val graph = Graph()
    private val insights = Insights()

    private val gson = GsonBuilder().setPrettyPrinting().create()

    fun extractFromText(text: String): Map<String, Any> {
        val cleanedText = cleaner.clean(text)
        val chunkList = chunker.chunk(cleanedText)
        val entityList = entities.extract(cleanedText)
        val relationList = relations.extract(entityList)
        val graphData = graph.build(entityList)
        val insightsData = insights.compute(entityList, chunkList, cleanedText)

        return buildResult("", "", cleanedText, chunkList, entityList, relationList, graphData, insightsData)
    }

    private fun buildResult(
        url: String,
        title: String,
        text: String,
        chunks: List<Chunk>,
        entityList: List<Entity>,
        relationList: List<Relation>,
        graphData: GraphData,
        insightsData: InsightsData
    ): Map<String, Any> {
        val sortedChunks = chunks.sortedBy { it.index }

        val sortedEntities = entityList.distinct().sortedWith(
            compareBy<Entity> { it.type }.thenBy { it.value }
        )

        val sortedRelations = relationList.sortedWith(
            compareBy<Relation> { it.source }.thenBy { it.target }
        )

        val sortedNodes = graphData.nodes.sortedBy { it.id }
        val sortedEdges = graphData.edges.sortedWith(
            compareBy<GraphEdge> { it.source }.thenBy { it.target }
        )

        val meta = linkedMapOf(
            "title" to title,
            "url" to url
        )

        val content = linkedMapOf(
            "text" to text
        )

        val chunksList = sortedChunks.map { chunk ->
            linkedMapOf(
                "text" to chunk.text,
                "index" to chunk.index,
                "start" to chunk.start,
                "end" to chunk.end
            )
        }

        val entitiesList = sortedEntities.map { entity ->
            linkedMapOf(
                "type" to entity.type,
                "value" to entity.value
            )
        }

        val relationsList = sortedRelations.map { relation ->
            linkedMapOf(
                "source" to relation.source,
                "target" to relation.target,
                "type" to relation.type
            )
        }

        val graphMap = linkedMapOf(
            "nodes" to sortedNodes.map { node ->
                linkedMapOf(
                    "id" to node.id,
                    "type" to node.type,
                    "value" to node.value
                )
            },
            "edges" to sortedEdges.map { edge ->
                linkedMapOf(
                    "source" to edge.source,
                    "target" to edge.target,
                    "weight" to edge.weight
                )
            }
        )

        val insightsMap = linkedMapOf(
            "top_entities" to insightsData.top_entities,
            "stats" to insightsData.stats,
            "entity_counts" to insightsData.entity_counts
        )

        return linkedMapOf(
            "meta" to meta,
            "content" to content,
            "chunks" to chunksList,
            "entities" to entitiesList,
            "relations" to relationsList,
            "graph" to graphMap,
            "insights" to insightsMap
        )
    }

    fun toJson(map: Map<String, Any>): String {
        return gson.toJson(map)
    }
}
